src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,706 @@
+    this->gotLen += to_read;
+
+    // Move pointer to the next record on the page
+    this->pageOffset += MAXALIGN(this->totLen);
+
+    // Rest of the record data is on the next page
+    while (this->gotLen != this->record->xl_tot_len)
+    {
+stepReadBody:
+        if (!readPage(this, input, stepReadBody))
+        {
+            return ReadRecordNeedBuffer;
+        }
+
+        if (this->currentHeader->xlp_info & XLP_FIRST_IS_OVERWRITE_CONTRECORD)
+        {
+            // This record has been overwritten.
+            // Write to the output what we managed to read as is, skipping filtering.
+            return ReadRecordSuccess;
+        }
+
+        if (!(this->currentHeader->xlp_info & XLP_FIRST_IS_CONTRECORD))
+        {
+            THROW_FMT(FormatError, "%s - should be XLP_FIRST_IS_CONTRECORD", strZ(pgLsnToStr(this->recPtr)));
+        }
+
+        if (this->currentHeader->xlp_rem_len == 0 ||
+            this->totLen != (this->currentHeader->xlp_rem_len + this->gotLen))
+        {
+            THROW_FMT(FormatError, "%s - invalid contrecord length: expect: %zu, get %u", strZ(pgLsnToStr(this->recPtr)),
+                      this->record->xl_tot_len - this->gotLen, this->currentHeader->xlp_rem_len);
+        }
+
+        size_t to_write = Min(this->currentHeader->xlp_rem_len, this->walPageSize - this->pageOffset);
+        memcpy(((char *) this->record) + this->gotLen, this->page + this->pageOffset, to_write);
+        this->pageOffset += MAXALIGN(to_write);
+        this->gotLen += to_write;
+    }
+    this->walInterface->validXLogRecord(this->record, this->heapPageSize);
+
+    if (this->record->xl_rmid == RM_XLOG_ID && this->record->xl_info == XLOG_SWITCH)
+    {
+        this->isSwitchWal = true;
+    }
+    this->recordNum++;
+    return ReadRecordSuccess;
+}
+
+static void
+filterRecord(WalFilterState *const this)
+{
+    // In the case of overwrite contrecord, we do not need to try to filter it, since the record may not have a body at all.
+    if (this->gotLen != this->record->xl_tot_len)
+    {
+        return;
+    }
+
+    const RelFileNode *node = getRelFileNodeGPDB6(this->record);
+    if (!node)
+    {
+        return;
+    }
+
+    bool isPassTheFilter = false;
+
+    // isRelationNeeded can allocate memory on the first call. Therefore, we switch the context.
+    MEM_CONTEXT_OBJ_BEGIN(this)
+    isPassTheFilter = isRelationNeeded(node->dbNode, node->spcNode, node->relNode);
+    MEM_CONTEXT_OBJ_END();
+
+    if (isPassTheFilter)
+    {
+        return;
+    }
+
+    this->record->xl_rmid = RM_XLOG_ID;
+    this->record->xl_info = XLOG_NOOP;
+    uint32_t crc = crc32cInit();
+    crc = crc32cComp(crc, (unsigned char *) XLogRecGetData(this->record), this->record->xl_len);
+    crc = crc32cComp(crc, (unsigned char *) this->record, offsetof(XLogRecord, xl_crc));
+    crc = crc32cFinish(crc);
+
+    this->record->xl_crc = crc;
```

Comment:
Is it possible to separate this code and similar code from the `hrnGpdbCreateXRecord` function into a separate function?